indexd_ffi/src/lib.rs

```diff
@@ -712,103 +687,125 @@ impl From<indexd::app_client::Account> for Account {
     }
 }
 
-/// Provides options for an upload operation.
-#[derive(uniffi::Record)]
-pub struct UploadOptions {
-    #[uniffi(default = 10)]
-    pub max_inflight: u8,
-    #[uniffi(default = 10)]
-    pub data_shards: u8,
-    #[uniffi(default = 20)]
-    pub parity_shards: u8,
-
-    /// Optional metadata to attach to the object.
-    /// This will be encrypted with the object's master key.
-    #[uniffi(default = None)]
-    pub metadata: Option<Vec<u8>>,
-    /// Optional callback to report upload progress.
-    /// The callback will be called with the number of bytes uploaded
-    /// and the total encoded size of the upload.
-    #[uniffi(default = None)]
-    pub progress_callback: Option<Arc<dyn UploadProgressCallback>>,
+#[derive(Debug, Error, uniffi::Error)]
+#[uniffi(flat_error)]
+pub enum ConnectError {
+    #[error("not connected")]
+    NotConnected,
+    #[error("already connected")]
+    AlreadyConnected,
+    #[error("app client error: {0}")]
+    AppClient(#[from] app_client::Error),
+    #[error("task error: {0}")]
+    JoinError(#[from] tokio::task::JoinError),
+    #[error("crypto error: {0}")]
+    Crypto(String),
+    #[error("host client error: {0}")]
+    Hosts(#[from] quic::Error),
+    #[error("error: {0}")]
+    Custom(String),
 }
 
-/// Provides options for a download operation.
+/// Metadata about an application connecting to the indexer.
 #[derive(uniffi::Record)]
-pub struct DownloadOptions {
-    #[uniffi(default = 10)]
-    pub max_inflight: u8,
-    #[uniffi(default = 0)]
-    pub offset: u64,
-    #[uniffi(default = None)]
-    pub length: Option<u64>,
+pub struct AppConnectOptions {
+    pub indexer_url: String,
+    pub app_key: Arc<AppKey>,
+
+    pub name: String,
+    pub description: String,
+    pub service_url: String,
+    pub logo_url: Option<String>,
+    pub callback_url: Option<String>,
 }
 
-/// An SDK enables interaction with an indexer and
-/// storage providers on the Sia network.
 #[derive(uniffi::Object)]
-pub struct SDK {
+pub struct UnauthenticatedSDK {
```

Comment:
How about using a type state here instead of 2 types? It's odd to have to call `UnauthenticatedSDK::new` rather than a `SDK::new` constructor

Reply:
In that case I'll leave it up to you if you want to keep this change or not. It's a lot less appealing to me this way.